tinygrad/codegen/linearizer.py

```diff
@@ -101,7 +101,7 @@ def global_load(self, i:int, idxs:List[Node], acc:Optional[LazyOp]=None, barrier
       key = f"{acc is not None}{localtype}{'CONST'+str(this_const) if this_const is not None and acc is None else (buf.idx if isinstance(buf, MemBuffer) else cast(LocalBuffer, buf).name)}{idx.render()}{valid.render()}"  # noqa: E501
       if key not in self.load_cache:
         if acc is not None:
-          self.load_cache[key] = self.uops.add(UOps.DEFINE_ACC, localtype, (), (self.get_reduce_acc(acc), i, acc_count))
+          self.load_cache[key] = self.uops.add(UOps.DEFINE_ACC, localtype, ((barrier,) if barrier else ()), (self.get_reduce_acc(acc), i, acc_count))
```

Comment:
so I think a couple things are wrong here:
1. load_cache[key] needs to update. acc is not None isn't sufficient if you have multiple reduces.
I tried doing this - You need to make it unique.
https://github.com/Qazalin/tinygrad/commit/bb677d6b106e29acef712cde02fd426cf940ebba
2. all uops are cachable now - i is the same for both reduces so it'll cache DEFINE_ACC, it shouldn't.

Reply:
the barrier is to prevent it from using a previous reduce from `saved_expr` or `load_cache[key]` but I think I found a better way that's more like https://github.com/Qazalin/tinygrad/commit/bb677d6b106e29acef712cde02fd426cf940ebba

w/o barrier:
```
__kernel void r_128_64n1(__global float* data0, const __global float* data1) {
  int gidx0 = get_group_id(0); /* 128 */
  float acc0 = 0.0f;
  int alu0 = (gidx0*64);
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
  for (int ridx1 = 0; ridx1 < 64; ridx1++) {
    float val1 = data1[alu0+ridx1];
    acc0 = ((val1-acc0)+acc0);
  }
  data0[gidx0] = acc0;
}
```

w/ barrier:
```
__kernel void r_128_64n1(__global float* data0, const __global float* data1) {
  int gidx0 = get_group_id(0); /* 128 */
  float acc0 = 0.0f;
  int alu0 = (gidx0*64);
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
  float acc1 = 0.0f;
  for (int ridx1 = 0; ridx1 < 64; ridx1++) {
    float val1 = data1[alu0+ridx1];
    acc1 = ((val1-acc0)+acc1);
  }
  data0[gidx0] = acc1;
}
```

Reply:
here: https://github.com/tinygrad/tinygrad/pull/4259/commits/c1f4cd539e0d4ece02e396d7a40d70abf30e9ed6#r1602269166